gensim/models/doc2vec.py

```diff
@@ -808,6 +809,50 @@ def delete_temporary_training_data(self, keep_doctags_vectors=True, keep_inferen
         if self.docvecs and hasattr(self.docvecs, 'doctag_syn0_lockf'):
             del self.docvecs.doctag_syn0_lockf
 
+    def save_word2vec_format(self, fname, doc_vec=True, word_vec=False, prefix='dt_', fvocab=None, binary=False):
+        """
+        Store the input-hidden weight matrix.
+
+         `fname` is the file used to save the vectors in
+         `doc_vec` is an optional boolean indicating whether to store document vectors
+         `word_vec` is an optional boolean indicating whether to store word vectors
+         (if both doc_vec and word_vec are True, then both vectors are stored in the same file)
+         `prefix` to uniquely indentify doctags from word vocab, and avoid collision
+         in case of repeated string in doctag and word vocab
+         `fvocab` is an optional file used to save the vocabulary
+         `binary` is an optional boolean indicating whether the data is to be saved
+         in binary word2vec format (default: False)
+
+        """
+        total_vec = len(self.docvecs)
+        if word_vec:
+            total_vec = total_vec + len(self.wv.vocab)
+        # save document vectors
+        if doc_vec:
+            logger.info("storing %sx%s projection weights into %s" % (total_vec, self.vector_size, fname))
+            with utils.smart_open(fname, 'wb') as fout:
+                fout.write(utils.to_utf8("%s %s\n" % (total_vec, self.vector_size)))
+                # store as in input order
+                for i in range(len(self.docvecs)):
+                    doctag = prefix + str(self.docvecs.index_to_doctag(i))
+                    row = self.docvecs.doctag_syn0[i]
+                    if binary:
+                        fout.write(utils.to_utf8(doctag) + b" " + row.tostring())
+                    else:
+                        fout.write(utils.to_utf8("%s %s\n" % (doctag, ' '.join("%f" % val for val in row))))
+            # save word vectors
+            if word_vec:
+                with utils.smart_open(fname, 'ab') as fout:
+                    for word, vocab in sorted(iteritems(self.wv.vocab), key=lambda item: -item[1].count):
+                        row = self.wv.syn0[vocab.index]
+                        if binary:
+                            fout.write(utils.to_utf8(word) + b" " + row.tostring())
+                        else:
+                            fout.write(utils.to_utf8("%s %s\n" % (word, ' '.join("%f" % val for val in row))))
+        else:
+            KeyedVectors.save_word2vec_format(self.wv, fname, fvocab=fvocab, binary=binary)
```

Comment:
A thought, in combination with my other comments: perhaps code duplication can be reduced by, if word-vecs are enabled, just calling KeyedVectors on the word-vectors first, *then* appending doc-vecs if necessary. This would reuse the word-writing (and vocab-writing) code from KeyedVectors, but then re-open the file for append to add doc-vectors (if enabled). To make sure the front-of-file vector-count was correct, the `KeyedVectors.load_word2vec_format()` would need a new parameter telling it to boost the count by some factor the caller knows.